### ISSUE TITLE
Race condition caused by dataplaneStateChange event results in invalid reservation state 

### ISSUE THREAD

**Issue by jmacauley** [esnet/sense-rm#22]
As the log segment below shows, when a reservation is terminated a dataplaneStateChange arrives just before the terminateConfirmed.  The dataplaneStateChange handling code reads the reservation record so it could update the dataplane state, but before it can be written back to the database, the terminateConfirmed handling code updates the reservation state to terminated.  Unfortunately, the dataplaneStateChange handling code now has an old version of the reservation state machine, so when it writes it back to the database we have an incorrect state machine.

1. Confirm this is the case.
2. Consider how to enforce proper sequencing, perhaps a separate table for the dataplaneState?
3. Did the model eventually correct itself on an audit?

The SENSE-O delta id that created the reservation is 8eee9038-d880-4c50-ba83-a55f95e217a1.

This issue also occurs on the provision and release operations.

```
021-11-04 07:20:47,903 [qtp65586123-3329] INFO  n.es.sense.rm.driver.nsi.NsiDriver (354) - [propagateDelta] processing deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc, modelId = d9feb30f-a6f4-4f2f-80e8-cea123f93b63
2021-11-04 07:20:47,906 [qtp65586123-3329] INFO  n.e.s.r.d.nsi.db.ModelServiceBean (112) - [ModelServiceBean] getCurrent() topologyId = urn:ogf:network:surf.nl:2020:production
2021-11-04 07:20:47,928 [qtp65586123-3329] INFO  n.es.sense.rm.driver.nsi.NsiDriver (407) - [NsiDriver] stored deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:47,928 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (218) - [processDelta] start deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:47,928 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (225) - [processDelta] processing reduction, deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:47,928 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (250) - [processDeltaReduction] start
2021-11-04 07:20:47,930 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (269) - [processDeltaReduction] SwitchingSubnet: urn:ogf:network:surf.nl:2020:production:switch:EVTS.A-GOLE:conn+8eee9038-d880-4c50-ba83-a55f95e217a1:resource+links-Connection_1:vlan+2025
2021-11-04 07:20:47,934 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (292) - [processDeltaReduction] done
2021-11-04 07:20:47,934 [qtp65586123-3329] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (237) - [processDelta] end deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:47,942 [qtp65586123-3329] INFO  n.e.s.r.m.MeasurementController (50) - { id: "62651fdb-8923-499f-aa42-3be33af8464f"
  measurement: "DELTA_RESERVE"
  resource: "ae6ec665-41b1-4855-8c79-9f9de63abddc"
  generated: "1636035647941"
  mtype: "duration"
  mvalue: "38"}
2021-11-04 07:20:47,943 [qtp65586123-3329] INFO  n.es.sense.rm.api.SenseRmController (1956) - [SenseRmController] Delta id = ae6ec665-41b1-4855-8c79-9f9de63abddc, lastModified = 2021-11-04T07:20:47.935-07:00, content-location = https://sense-rm.es.net/netherlight/api/sense/v1/deltas/ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:47,943 [qtp65586123-3329] INFO  n.es.sense.rm.api.SenseRmController (1979) - [SenseRmController] Delta returning id = ae6ec665-41b1-4855-8c79-9f9de63abddc, creationTime = 2021-11-04T07:20:47.935-07:00
2021-11-04 07:20:48,206 [main] INFO  net.es.sense.rm.Application (59) - [SENSE-N-RM] Memory use :206M/966M
2021-11-04 07:20:48,682 [qtp65586123-2087] INFO  n.es.sense.rm.api.SenseRmController (2095) - [SenseRmController] operation = https://sense-rm.es.net/netherlight/api/sense/v1/deltas/ae6ec665-41b1-4855-8c79-9f9de63abddc/actions/commit, deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:48,682 [qtp65586123-2087] INFO  n.es.sense.rm.driver.nsi.NsiDriver (467) - [NsiDriver] processing commitDelta for id = ae6ec665-41b1-4855-8c79-9f9de63abddc
2021-11-04 07:20:48,685 [qtp65586123-2087] INFO  n.es.sense.rm.driver.nsi.NsiDriver (482) - [NsiDriver] delta state, id = ae6ec665-41b1-4855-8c79-9f9de63abddc, state = Accepted.
2021-11-04 07:20:48,691 [qtp65586123-2087] INFO  o.a.c.w.s.f.ReflectionServiceFactoryBean (437) - Creating Service {http://schemas.ogf.org/nsi/2013/12/connection/provider}ConnectionProviderPortService from class org.ogf.schemas.nsi._2013._12.connection.provider.ConnectionProviderPort
2021-11-04 07:20:49,725 [qtp65586123-2087] DEBUG n.e.s.rm.driver.nsi.cs.CsProvider (691) - [csProvider] issued terminate operation correlationId = urn:uuid:fa50adf0-d425-45d0-8da9-1b396fe36fbe, connectionId = e7dd8b45-3f01-4f5d-af6d-7bb894c07056
2021-11-04 07:20:49,725 [qtp65586123-2087] INFO  n.e.s.rm.driver.nsi.cs.CsProvider (953) - [CsProvider] waiting for completion of correlationId = urn:uuid:fa50adf0-d425-45d0-8da9-1b396fe36fbe
2021-11-04 07:20:58,206 [main] INFO  net.es.sense.rm.Application (59) - [SENSE-N-RM] Memory use :292M/966M
2021-11-04 07:21:08,206 [main] INFO  net.es.sense.rm.Application (59) - [SENSE-N-RM] Memory use :292M/966M
2021-11-04 07:21:08,667 [qtp65586123-4307] INFO  n.e.s.r.d.n.c.api.ConnectionService (596) - [ConnectionService] terminateConfirmed received for correlationId = urn:uuid:fa50adf0-d425-45d0-8da9-1b396fe36fbe, connectionId: e7dd8b45-3f01-4f5d-af6d-7bb894c07056


2021-11-04 07:21:08,668 [qtp65586123-2165] INFO  n.e.s.r.d.n.c.api.ConnectionService (923) - [ConnectionService] dataPlaneStateChange for providerNSA = urn:ogf:network:netherlight.net:2013:nsa:safnari, correlationId = urn:uuid:d0448b40-396a-480a-a266-1be3da5b29ed, dataPlaneStateChange:
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns2:dataPlaneStateChange xmlns:ns6="http://schemas.ogf.org/nsi/2013/12/framework/types" xmlns:ns5="http://www.w3.org/2001/04/xmlenc#" xmlns:ns8="http://schemas.ogf.org/nsi/2013/12/framework/headers" xmlns:ns7="http://schemas.ogf.org/nsi/2013/12/services/point2point" xmlns:ns2="http://schemas.ogf.org/nsi/2013/12/connection/types" xmlns:ns4="http://www.w3.org/2000/09/xmldsig#" xmlns:ns3="urn:oasis:names:tc:SAML:2.0:assertion">
    <connectionId>e7dd8b45-3f01-4f5d-af6d-7bb894c07056</connectionId>
    <notificationId>2</notificationId>
    <timeStamp>2021-11-04T14:21:07.050112Z</timeStamp>
    <dataPlaneStatus>
        <active>false</active>
        <version>0</version>
        <versionConsistent>true</versionConsistent>
    </dataPlaneStatus>
</ns2:dataPlaneStateChange>

2021-11-04 07:21:08,669 [qtp65586123-4307] INFO  n.e.s.r.d.n.c.api.ConnectionService (609) - [ConnectionService] terminateConfirmed: storing reservation update, cid = e7dd8b45-3f01-4f5d-af6d-7bb894c07056

2021-11-04 07:21:08,670 [qtp65586123-2165] INFO  n.e.s.r.d.n.c.api.ConnectionService (947) - [ConnectionService] updating connectionId = e7dd8b45-3f01-4f5d-af6d-7bb894c07056, reservation Reservation{id=689, discovered=1636035477201, providerNsa='urn:ogf:network:netherlight.net:2013:nsa:safnari', globalReservationId='urn:ogf:network:surf.nl:2020:production:switch:EVTS.A-GOLE:conn+8eee9038-d880-4c50-ba83-a55f95e217a1:resource+links-Connection_1:vlan+2025', description='deltaId+387b66ef-dd75-44c2-8b4f-b105748a7dc4:uuid+e1c0d4bb-d38d-4875-8572-6e5bba917f4e', parentConnectionId='null', connectionId='e7dd8b45-3f01-4f5d-af6d-7bb894c07056', topologyId='urn:ogf:network:surf.nl:2020:production', serviceType='http://services.ogf.org/nsi/2013/12/descriptions/EVTS.A-GOLE', startTime=0, endTime=1667571434071, reservationState=RESERVE_START, provisionState=RELEASED, lifecycleState=CREATED, dataPlaneActive=true, version=0, errorState=NONE, errorMessage=null, service='<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns7:p2ps xmlns:ns6="http://schemas.ogf.org/nsi/2013/12/framework/types" xmlns:ns5="http://www.w3.org/2001/04/xmlenc#" xmlns:ns8="http://schemas.ogf.org/nsi/2013/12/framework/headers" xmlns:ns7="http://schemas.ogf.org/nsi/2013/12/services/point2point" xmlns:ns2="http://schemas.ogf.org/nsi/2013/12/connection/types" xmlns:ns4="http://www.w3.org/2000/09/xmldsig#" xmlns:ns3="urn:oasis:names:tc:SAML:2.0:assertion">
    <capacity>1000</capacity>
    <directionality>Bidirectional</directionality>
    <symmetricPath>true</symmetricPath>
    <sourceSTP>urn:ogf:network:surf.nl:2020:production:netherlight.cern-1?vlan=2025</sourceSTP>
    <destSTP>urn:ogf:network:surf.nl:2020:production:netherlight.canarie-1?vlan=2025</destSTP>
</ns7:p2ps>
'}

2021-11-04 07:21:08,670 [qtp65586123-2165] INFO  n.e.s.r.d.n.c.api.ConnectionService (950) - [ConnectionService] writing updated connectionId = e7dd8b45-3f01-4f5d-af6d-7bb894c07056, reservation Reservation{id=689, discovered=1636035668670, providerNsa='urn:ogf:network:netherlight.net:2013:nsa:safnari', globalReservationId='urn:ogf:network:surf.nl:2020:production:switch:EVTS.A-GOLE:conn+8eee9038-d880-4c50-ba83-a55f95e217a1:resource+links-Connection_1:vlan+2025', description='deltaId+387b66ef-dd75-44c2-8b4f-b105748a7dc4:uuid+e1c0d4bb-d38d-4875-8572-6e5bba917f4e', parentConnectionId='null', connectionId='e7dd8b45-3f01-4f5d-af6d-7bb894c07056', topologyId='urn:ogf:network:surf.nl:2020:production', serviceType='http://services.ogf.org/nsi/2013/12/descriptions/EVTS.A-GOLE', startTime=0, endTime=1667571434071, reservationState=RESERVE_START, provisionState=RELEASED, lifecycleState=CREATED, dataPlaneActive=false, version=0, errorState=NONE, errorMessage=null, service='<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<ns7:p2ps xmlns:ns6="http://schemas.ogf.org/nsi/2013/12/framework/types" xmlns:ns5="http://www.w3.org/2001/04/xmlenc#" xmlns:ns8="http://schemas.ogf.org/nsi/2013/12/framework/headers" xmlns:ns7="http://schemas.ogf.org/nsi/2013/12/services/point2point" xmlns:ns2="http://schemas.ogf.org/nsi/2013/12/connection/types" xmlns:ns4="http://www.w3.org/2000/09/xmldsig#" xmlns:ns3="urn:oasis:names:tc:SAML:2.0:assertion">
    <capacity>1000</capacity>
    <directionality>Bidirectional</directionality>
    <symmetricPath>true</symmetricPath>
    <sourceSTP>urn:ogf:network:surf.nl:2020:production:netherlight.cern-1?vlan=2025</sourceSTP>
    <destSTP>urn:ogf:network:surf.nl:2020:production:netherlight.canarie-1?vlan=2025</destSTP>
</ns7:p2ps>
'}


2021-11-04 07:21:08,673 [qtp65586123-2087] INFO  n.e.s.rm.driver.nsi.cs.CsProvider (957) - [CsProvider] operation terminate completed, correlationId = urn:uuid:fa50adf0-d425-45d0-8da9-1b396fe36fbe
2021-11-04 07:21:08,674 [qtp65586123-2087] INFO  n.es.sense.rm.driver.nsi.NsiDriver (504) - [NsiDriver] delta state, id = ae6ec665-41b1-4855-8c79-9f9de63abddc, state = Committing.
2021-11-04 07:21:08,674 [qtp65586123-2087] INFO  n.es.sense.rm.driver.nsi.NsiDriver (507) - [NsiDriver] delta state transition to id = ae6ec665-41b1-4855-8c79-9f9de63abddc, state = Committed.


2021-11-04 07:21:08,679 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] INFO  n.e.s.r.driver.nsi.AuditServiceBean (77) - [AuditService] starting audit for urn:ogf:network:surf.nl:2020:production.
2021-11-04 07:21:08,679 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] INFO  n.e.s.r.driver.nsi.AuditServiceBean (78) - [AuditService] lastDds = 1636035239000, lastCon = 1636035477201
2021-11-04 07:21:08,680 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.r.d.n.c.d.ReservationServiceBean (30) - ReservationServiceBean: getLastDiscovered returned 1636035668670
2021-11-04 07:21:08,680 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.r.driver.nsi.AuditServiceBean (92) - [AuditService] generating NML topology model, new dds = 1636035239000, con = 1636035668670
2021-11-04 07:21:08,680 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.r.driver.nsi.AuditServiceBean (101) - [AuditService] processing topologyId = urn:ogf:network:surf.nl:2020:production
2021-11-04 07:21:08,681 [qtp65586123-2087] INFO  n.e.s.r.m.MeasurementController (50) - { id: "e6adf7bb-92c6-4813-86c7-ef5d00411fbe"
  measurement: "DELTA_COMMIT"
  resource: "ae6ec665-41b1-4855-8c79-9f9de63abddc"
  generated: "1636035668680"
  mtype: "duration"
  mvalue: "19998"}
2021-11-04 07:21:08,682 [qtp65586123-2087] INFO  n.es.sense.rm.api.SenseRmController (2120) - [SenseRmController] commitDelta deltaId = ae6ec665-41b1-4855-8c79-9f9de63abddc, lastModified = 2021-11-04T07:21:08.674-07:00


2021-11-04 07:21:08,708 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.r.d.n.m.SwitchingSubnetModel (91) - [SwitchingSubnetModel] loading...
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (774) - [getSwitchingServicesIndexed] entering.
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (780) - [getSwitchingServicesIndexed] relation = http://schemas.ogf.org/nml/2013/05/base#hasInboundPort
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (780) - [getSwitchingServicesIndexed] relation = http://schemas.ogf.org/nml/2013/05/base#hasOutboundPort
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (780) - [getSwitchingServicesIndexed] relation = http://schemas.ogf.org/nml/2013/05/base#hasService
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (784) - [getSwitchingServicesIndexed] SwitchingService instance net.es.nsi.dds.lib.jaxb.nml.NmlSwitchingServiceType
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (787) - found SwitchingService urn:ogf:network:surf.nl:2020:production:switch:EVTS.A-GOLE
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (803) - SwitchingService found but no port members so adding matching, topologyId = urn:ogf:network:surf.nl:2020:production
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (809) - Adding ports to SwitchingService = urn:ogf:network:surf.nl:2020:production:switch:EVTS.A-GOLE
2021-11-04 07:21:08,735 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (822) - [getSwitchingServicesIndexed] no ports defined so populating wildcard.
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:staging.automation.surf.net:2017:staging
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:daej-nsi.kreonet.net:2020:daej_topo2
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:daej-nsi.kreonet.net:2020:daej_topo1
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:lsanca.pacificwave.net:2016:mx204-2
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:lsanca.pacificwave.net:2016:losa2-pw-sdn-1
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:sinet.ac.jp:2013:topology
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:icair.org:2013:slex9208
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:sl9208.icair.org:2013:topology
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:uhnet.net:2021:hawaii-scidmz
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:jgn-x.jp:2013:topology
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:southernlight.net.br:2013:topology
2021-11-04 07:21:09,929 [SENSE-NSI-System-akka.actor.default-dispatcher-4078] DEBUG n.e.s.rm.driver.nsi.mrml.NmlModel (90) - [NmlModel] processing NML model urn:ogf:network:surf.nl:2020:production
```
[sense-n-rm.log.gz](https://github.com/esnet/sense-rm/files/7479235/sense-n-rm.log.gz)

**Comment by jmacauley** [comment]
Closed with commit 2d31618aeb9f25b2620b54795e753b00f7aed5f3.

**Comment by jmacauley** [comment]
This issue was solved on Nov 11th by update [2d31618](https://github.com/esnet/sense-rm/commit/2d31618aeb9f25b2620b54795e753b00f7aed5f3).
